test/integration/apiserver/admissionwebhook/match_conditions_test.go

```diff
@@ -0,0 +1,659 @@
+package admissionwebhook
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"io"
+	"net/http"
+	"net/http/httptest"
+	"sync"
+	"testing"
+	"time"
+
+	clientset "k8s.io/client-go/kubernetes"
+
+	admissionv1 "k8s.io/api/admission/v1"
+	admissionregistrationv1 "k8s.io/api/admissionregistration/v1"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	genericfeatures "k8s.io/apiserver/pkg/features"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	featuregatetesting "k8s.io/component-base/featuregate/testing"
+	apiservertesting "k8s.io/kubernetes/cmd/kube-apiserver/app/testing"
+	"k8s.io/kubernetes/test/integration/framework"
+)
+
+type admissionRecorder struct {
```

Comment:
Alternative approach I was considering when writing https://github.com/ivelichkovich/kubernetes/pull/4 was to have the validation policy reject by default and then assert that matched pods don't exist. This seemed more full-proof though  

For mutating, we can have the webhook patch some metadata (label or annotation) and check that only matched pods have those labels.

Reply:
Yeah I like this way, thinking to do the same for mutating and just record and do a no-op mutation